### ISSUE TITLE
Converting Specific JXL -> PNG with convert Breaks Colors

### ISSUE THREAD

**Issue by gianni-rosato** [ImageMagick/ImageMagick#6157]
### ImageMagick version

7.1.1-2

### Operating system

Linux

### Operating system, version and so on

Linux x86_64 6.2.5-zen1-1-zen

### Description

Upon trying to convert my JXL image to a PNG with `convert` (the JXL image was re-encoded from a 16-bit PNG exported from Darktable) it will cause the output image to be discolored.

This happens on a number of Darktable output images I have, which is a big problem.

### Steps to Reproduce

1. Download `smalltest_source` linked below
2. Reencode with *any version* of `cjxl`
3. run `convert output.jxl output.png`
4. Observe discolored PNG. You may open in a web browser like Firefox to see the behavior I'm talking about.

### Images

![smalltest_source](https://user-images.githubusercontent.com/35711760/224607920-ca04044d-cf02-49cf-80c3-115eb1d95d71.png)
![jxl_file](https://cdn.discordapp.com/attachments/794206170445119489/1084693938532782170/smalltest_enc.jxl)
![convert_output](https://user-images.githubusercontent.com/35711760/224608314-81853436-98e2-4f2f-a163-7fa6621ec277.png)

**Comment by urban-warrior** [comment]
Your image is linear RGB.  Try try this command to convert it to sRGB:

```
convert output.jxl -colorspace sRGB output.png
```

**Comment by gianni-rosato** [comment]
Is there any way to do this while maintaining the AdobeRGB colorspace?

**Comment by fmw42** [comment]
Did you try

`magick output.jxl -colorspace AdobeRGB output.png`

or whatever your AdobeRGB profile name actually is.

If on IM 7, I suggest you use magick not convert.

**Comment by gianni-rosato** [comment]
`magick: unrecognized colorspace '-colorspace' 'AdobeRGB' at CLI arg 2 @ error/operation.c/CLISettingOptionInfo/639.`

I must need to find the path needed to access it. For now, converting to sRGB is gonna be fine for me. Thank you for your help!

**Comment by kmilos** [comment]
> Is there any way to do this while maintaining the AdobeRGB colorspace?

Did you try the "keep original" color space option when exporting from darktable? The JPEG XL file will be considerably larger, but it'll preserve the profile. Otherwise, it discards it when encoding and only decodes to linear RGB (unless you're doing lossless, i.e. quality 100 as well when you get out what you put in).

After that, I don't know if IM will recognize and preserve the color space of such JPEG XL files without these extra options...

**Comment by snibgo** [comment]
There is some confusion on this thread.

`-colorspace` should be used with a keyword, not a filename. The keyword might be `Adobe98` or `sRGB` among others, but not `AdobeRGB`. Use `magick -list colorspace` for a complete list.

`-profile` should be used with the name of an ICC color profile file. For example `sRGB.icc` or `AdobeRGB1998.icc`. Give the path to the file if it isn't in the current directory.

The file `smalltest_enc.jxl` has an embedded ICC profile. This defines a non-linear tone response curve (TRC). However, the pixel data seems to be linear, without a TRC. According to https://en.wikipedia.org/wiki/Adobe_RGB_color_space , "The precise gamma value is 563/256, or 2.19921875."

So we can change the pixel values like this:
```
magick smalltest_enc.jxl -evaluate Pow %[fx:256/563] out.tiff
```
The pixel data is now consistent with the embedded profile.

If we want to convert the image to sRGB, we should use `-profile`:
```
magick smalltest_enc.jxl -evaluate Pow %[fx:256/563] -profile sRGB.icc out2.tiff
```

**Comment by snibgo** [comment]
You may be correct about sRGB primaries; I don't know. If you are correct, then to the question above from gianni-rosato ...
> Is there any way to do this while maintaining the AdobeRGB colorspace?

... the answer is: the image has been returned encoded as sRGB (with a misleading AdobeRGB1998 profile). You can strip the misleading profile, then convert it to whatever colorspace you want.

**Comment by kmilos** [comment]
I wouldn't say the embedded profile is misleading, more that libjxl is complicated 😉

**Comment by kmilos** [comment]
I think there might be a bug here:

https://github.com/ImageMagick/ImageMagick/blob/main/coders/jxl.c#L463-L471

The decoded pixel data is returned in TARGET_DATA color space, bit IM gets and assigns TARGET_ORIGINAL ICC profile, while it should get and assign the TARGET_DATA ICC profile (or do the conversion of pixel data from TARGET_DATA back to TARGET_ORIGINAL).

**Comment by urban-warrior** [comment]
The issue is that the image is linear RGB.  We assign the linear color space but did not assign a gamma of 1.0.  With the patch, you can return sRGB pixels with:

```
magick smalltest_enc.jxl -colorspace sRGB out.tiff
```

**Comment by kmilos** [comment]
One thing about JPEG XL: there is no such thing as "image is linear RGB". The image is a JPEG XL stream, which internally uses XYB color space. What matters is what you get out of that stream after the decoding step. It is (currently) either sRGB for integer output or linear RGB for float (w/ sRGB primaries), no matter what the original color space of the image was _before_ encoding.

**Comment by urban-warrior** [comment]
Correction then, the transfer function is linear.

**Comment by kmilos** [comment]
The issue is that you assign the original UCC profile, but the JPEG XL stream is not decoded into that color space.

**Comment by ValZapod** [comment]
>Your image is linear RGB. Try try this command to convert it to sRGB:

Nope, it is not linear. First of all, it is R'G'B', because almost all JXL is RGB. But then looking into ffplay smalltest_enc.jxl it prints:

[libjxl @ 00000194dcd55bc0] Falling back on wide gamut output
[libjxl @ 00000194dcd55bc0] Falling back on iec61966-2-1/sRGB transfer


and looking into code we see that if (avctx->color_primaries == AVCOL_PRI_UNSPECIFIED) https://github.com/FFmpeg/FFmpeg/blob/c56f5be6782014fee165d361de1f548eaac7a272/libavcodec/libjxldec.c#L285

which apperently means not BT.709 primaries of sRGB, but in fact BT.2020 primaries.

Same for transfer: we are inside avctx->color_trc == AVCOL_TRC_UNSPECIFIED  (line 294) and then they go and check for FLOAT and that will mean linear trasnfer, but cause it is not FLOAT you get sRGB transfer. So I do not understand why you think it is linear. No transfer and no primaries are specified, which means it is BT.2020 primaries and sRGB transfer.

**Comment by urban-warrior** [comment]
The image color coding is JXL_COLOR_SPACE_RGB.  ImageMagick assigns it the RGB colorspace with a gamma of 1.0.  We then transform to the sRGB colorspace with a gamma of 0.454545 to get the image @gianni-rosato requested.   If that is not the correct workflow, let us know and we will correct it.

**Comment by ValZapod** [comment]
>If that is not the correct workflow, let us know and we will correct it.

The way it works you need to look into the spec's defaults if no primaries or transfer is specified.

**Comment by kmilos** [comment]
> The image color coding is JXL_COLOR_SPACE_RGB

This is not a color space definition. This just means 3 channels instead of e.g. grayscale.

**Comment by fmw42** [comment]
3 channels usually means RGB. But RGB can be linear (usually darker) or non-linear (normal view).

**Comment by ValZapod** [comment]
>RGB can be linear (usually darker) or non-linear (normal view).

In fact both look the same. If it does not it is a bug.

**Comment by kmilos** [comment]
> 3 channels usually means RGB

Again, what does this mean? RGB, but with what primaries?
